features/329-principal-apis/design.md

```diff
@@ -0,0 +1,100 @@
+provide a consistent API for manipulating principals.
+
+
+# Use Cases
+- Get which method is used to log in the current session
+- List all methods to authenticated as a user
+- Add/remove user authentication method
+- Send welcome email to OAuth accounts
+- Populate user metadata according to external profile
+- Create password credentials for users signed up with OAuth
+
+
+# Glossary
+
+**Identity Provider**:
+A method to assert the identity of user. For example, providing a password or
+perform OAuth flow.
+
+**Principal**:
+An identity associated with user. A user can have multiple principals. Since
+the term 'Principal' is defined differently across many frameworks and
+standards, we will instead use a simpler term 'Identity' for public interfaces.
+
+
+# Proposed Design
+
+## Re-authentication for Security
+Many principal manipulation operations are critical for security:
+- Add/remove login ID
+- Link/unlink SSO provider
+- Change password
+
+By default, user are required to re-authenticate before performing these
+operations. This is achieve by checking the issue time of the access token is
+recent.
+
+For change password, we support the common case of providing old password to
+implicitly authenticate the user.
+
+Both the re-auth requirement and time to considered 'recent' can be configured.
+Developer can turn off the re-auth requirement in configuration. However,
+developer should understand the potential security risk.
+
+## Current Principal
+We need to provide a way for developer to get the principal is associated with
+the current session:
+- Which login ID is used to login?
+- Which OAuth provider is used to login?
+
+When logging in using a principal, the ID of the principal (i.e. the current
+principal) would be embedded in the access token along with the user ID.
+Information about the current principal would be returned in user object.
+
+## Principal List
+We need to provide a way for developer to get a list of principals belonged to
+the user:
+- What login IDs the user has?
+- Which OAuth providers had the user linked?
+
+A list of principals belonged to current user would be returned through an API.
+
+# Creating Principal
+We need to provide a way for developer to create new principal for the user:
+- Add new login ID
+- Link OAuth account
+
+Each identity provider should have their specific API to create principal. APIs
+that create principals should have common behaviors:
+- Should respect re-authentication configuration
+
+# Deleting Principal
+We need to provide a way for developer to delete new principal for the user:
+- Remove login ID
+- Unlink OAuth account
+
+Each identity provider should have their specific API to delete principal. APIs
+that delete principals should have common behaviors:
+- Should respect re-authentication configuration
+- Should not allow a verified user to become unverified through the operation
+- Should not allow deleting the current principal
```

Comment:
Doesn't seems we should have this behaviour?

I'm thinking about the following use cases:

0. Context: A simple system that allow ONE login ID per user only
1. User `loginIDA` logged in, and want to change it's username
2. Since it can't rename a Login ID, the user need to delete the old login ID and add a new one.
3. Since it does not allow deleting the current principal, it also can't delete the old Login ID.
4. The only way to implement this is add the new username as new login ID, ask the user to login again, and delete the old login ID automatically.

The above behaviour seems to contradict with most "change my username"'s operation.

So either:
1. We allow rename of loginID
2. We allow delete of current principal.

Reply:
Both option is undesirable:
- allow renaming: we cannot ensure a principal is always referring the same identity
- allow deleting current principal: we cannot return a meaning current principal after the operation.

Instead, I'll add a `replaceLoginID` API, and practically it performs the suggested procedure (i.e. add new login ID then delete old login ID) atomically with necessary checks.

Reply:
Any reason why we should call it `replaceLoginID` instead of `renameLoginID` or `changeLoginID`? From usability perspective we should limits the usage of verb, so better check other verbs used in auth API and pick the common one. (We used to have `change` and `update` in different places and it was very confusing for users)

Reply:
'rename' or 'change' implies a mutable principal, which is not the case; this API creates a new principal and delete the old one as an atomic operation.

Reply:
I think consistency of verbs is more important, can we look at the list of verbs we use and use a common one for renaming of LoginID?

Actually I'm not sure what mutable meant here. In the above cases, if a logged in user changed the username (login ID), its identity's login ID is updated already? Even if you create a new auth token, for the end user it is still just like the identity's parameter was changed.

Reply:
It is not renaming login ID, it is replacing old login ID with new login ID. The ID of the current principal will be different.

Reply:
So it is still not `replaceLoginID`, it could be `replaceIdentity` but then it is very difficult to understand for newcomers.

For consistency and usability reason, it is better for us to follow the verbs of similar APIs in our platform like `changeLoginID` or `updateLoginID` or whatever we have been using for consistency.

Reply:
I'll name it `updateLoginID`. However, the documentation would need to note that the ID of identities would be changed; most user would not expect an 'update' operation would change the ID of an entity.

Reply:
You can note it as a side effect in spec, and we can mention it in documentations.

Reply:
Actually, what are the use cases for exposing the id of Identity?